### PR TITLE
[Minor]RadSite Performance Improve

### PR DESCRIPTION
**Pull request by NetsuNegi**
Added CellExt to store RadSite pointers to avoid constantly traversing RadSiteClass: Array.

### CODE REVIEW

File: src/Ext/Cell/Body.h [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+	using base_type = CellClass;
+
+	static constexpr DWORD Canary = 0xFDC49191;
+	static constexpr size_t ExtPointerOffset = 0x18;
+
+	class ExtData final : public Extension<CellClass>
+	{
+	public:
+		std::vector<RadSiteClass*> RadSites {};
+		std::vector<std::pair<RadSiteClass*, int>> RadLevels { };
```
[the remark this answers is not in the document]
Comment (NetsuNegi) [reply src/Ext/Cell/Body.h:20]: > `PhobosMap` is technically any reason to not use that instead ?`std::vector<std::pair<>>`

how to traverse `PhobosMap`？

**Comment by Coronia** [comment]
current test result: under environment with almost same amount of radiation and other conditions being as similar as possible, this pull request did increase performance a lot in all these conditions:
- no radiation on the map
- a bunch of units entering rad field
- radiation with RadHasInvoker/Owner
- radiation that can damage buildings

It's theoretically a bit laggier a the moment of radiation being created, but our test so far indicated that it's not worse than the current method
Hope others could give some more accurate test and check if all rad behaviors work properly

**Comment by NetsuNegi** [comment]
Due to unknown reasons, the S/L has been broken.
I dont't know how to fix it

**Comment by Metadorius** [comment]
Try uncommenting 0x144 offset, IIRC I had no trouble with it

**Comment by Metadorius** [comment]
from @Otamaa on Discord:

> iam going different route for optimization 
> which using `RadSiteClass::Ai` to deliver damage instead
> `cellExt` rad vector ,is probably used for adding/creating new rad
> no more cursed radsite array inside techno every techno

**Review by Starkku — CHANGES_REQUESTED** [review]
The radiation functionality appears to work (although there may be edge cases I did not catch in my tests) and the idea is sound, there is likely performance improvement to gain here.

However the implementation has some facets that likely need to be reworked. Main one is that `std::vector<std::pair<RadSiteClass*, int>>` is the cause of the save/load issues as whatever save/load functionality is implemented for vectors might not handle `std::pair` correctly. `PhobosMap` is effectively the same thing under the hood but it does not grant direct access to the underlying vectors or implement iteration functions, so switching to it would require some changes regardless.

There may be some alternate approaches to this like `CellExt` only tracking RadSites and RadSites themselves tracking levels per cell but I don't know what is the most feasible one right now.

**Comment by NetsuNegi** [comment]
> The radiation functionality appears to work (although there may be edge cases I did not catch in my tests) and the idea is sound, there is likely performance improvement to gain here.
> 
> However the implementation has some facets that likely need to be reworked. Main one is that `std::vector<std::pair<RadSiteClass*, int>>` is the cause of the save/load issues as whatever save/load functionality is implemented for vectors might not handle `std::pair` correctly. `PhobosMap` is effectively the same thing under the hood but it does not grant direct access to the underlying vectors or implement iteration functions, so switching to it would require some changes regardless.
> 
> There may be some alternate approaches to this like `CellExt` only tracking RadSites and RadSites themselves tracking levels per cell but I don't know what is the most feasible one right now.

I'll try breaking it down into two std: vector

**Comment by Starkku** [comment]
Otamaa's approach with all radiation logic being in RadSite and it keeping track which cells it exists on might be saner and cleaner in the end.

**Comment by NetsuNegi** [comment]
I‘m using a struct instead of std::pair, and it has own Save() Load() Serialize()
hope it will be useful

**Comment by Starkku** [comment]
I can confirm that S/L works now, not sure if any further tests with radiation functionality itself are required.

**Comment by NetsuNegi** [comment]
> I can confirm that S/L works now, not sure if any further tests with radiation functionality itself are required.

ok

**Comment by Coronia** [comment]
other logic for regular/custom radiation worked properly after the S/L fix

File: src/Ext/Bullet/Body.cpp [hunk 2]
```diff
@@ -94,28 +95,25 @@ void BulletExt::ExtData::InterceptBullet(TechnoClass* pSource, WeaponTypeClass*
...
-			if (pRadExt->Type != pRadType)
-				return false;
+	const auto pThis = this->OwnerObject();
 
-			if (MapClass::Instance.TryGetCellAt(pSite->BaseCell) != MapClass::Instance.TryGetCellAt(Cell))
-				return false;
+	const auto pWeapon = pThis->GetWeaponType();
+	const auto pWeaponExt = WeaponTypeExt::ExtMap.Find(pWeapon);
+	const auto pRadType = pWeaponExt->RadType;
+	const auto pThisHouse = pThis->Owner ? pThis->Owner->Owner : this->FirerHouse;
```
[the remark this answers is not in the document]
Comment (NetsuNegi) [reply src/Ext/Bullet/Body.cpp:108]: > This can be placed just before `RadSiteExt::CreateInstance`.

what?

File: src/Ext/RadSite/Hooks.cpp [hunk 3]
```diff
@@ -160,19 +165,14 @@ DEFINE_HOOK(0x43FB23, BuildingClass_AI_Radiation, 0x5)
...
-
 			if (pBuilding->IsAlive) // simple fix for previous issues
 			{
 				int damage = Game::F2I(radLevel * pType->GetLevelFactor());
 
 				if (maxDamageCount > 0)
 					damageCounts[pRadSite]++;
 
-				if (!pRadExt->ApplyRadiationDamage(pBuilding, damage, Game::F2I(orDistance)))
+				if (!pRadExt->ApplyRadiationDamage(pBuilding, damage, 0))
```
[the remark this answers is not in the document]
Comment (NetsuNegi) [reply src/Ext/RadSite/Hooks.cpp:175]: No, this parameter is required. The function does not define a default value for this parameter

**Review by CrimRecya — APPROVED** [review]

File: src/Ext/RadSite/Body.cpp [hunk 4]
```diff
@@ -60,6 +62,10 @@ void RadSiteExt::CreateInstance(CellStruct location, int spread, int amount, Wea
 	pRadSite->SetSpread(spread);
 	pRadExt->SetRadLevel(amount);
 	pRadExt->CreateLight();
+
+	const auto pCell = MapClass::Instance.TryGetCellAt(location);
```
Comment (Coronia) [inline src/Ext/RadSite/Body.cpp:66]: should there be a sanity check here since it's using TryGetCellAt?

Comment (NetsuNegi) [reply src/Ext/RadSite/Body.cpp:66]: Actually, there have been checks before, but I still added one

**Review by Coronia — APPROVED** [review]

File: src/Ext/RadSite/Hooks.cpp [hunk 5]
```diff
@@ -130,26 +129,32 @@ DEFINE_HOOK(0x43FB23, BuildingClass_AI_Radiation, 0x5)
...
 		CellStruct nCurrentCoord = buildingCoords + *pFoundation;
+		const auto pCell = MapClass::Instance.TryGetCellAt(nCurrentCoord);
+
+		if (!pCell)
+			continue;
+
+		const auto pCellExt = CellExt::ExtMap.Find(pCell);
 
-		for (auto const pRadSite : RadSiteClass::Array)
+		for (const auto& [pRadSite, radLevel] : pCellExt->RadLevels)
```
Comment (Metadorius) [inline src/Ext/RadSite/Hooks.cpp:145]: nice syntax, didn't know C++ has such 👍

File: src/Ext/RadSite/Hooks.cpp [hunk 6]
```diff
@@ -333,3 +327,96 @@ DEFINE_HOOK(0x65BB67, RadSite_Deactivate, 0x6)
...
+	const auto it = std::find_if(radLevels.begin(), radLevels.end(), [pThis](CellExt::RadLevel const& item) { return item.Rad == pThis; });
+
+	if (it != radLevels.end())
+	{
+		const int amount = Game::F2I(static_cast<double>(max - distance) / max * pThis->RadLevel / pThis->LevelSteps);
+		it->Level -= amount;
+	}
+
+	return SkipGameCode;
+}
```
Comment (Metadorius) [inline src/Ext/RadSite/Hooks.cpp:422]: Those 3 hooks seem to have a lot of repeated code, could you make a function or share the hook code using `DEFINE_HOOK_AGAIN` and checking `R->Origin()` address?

**Review by Metadorius — CHANGES_REQUESTED** [review]

File: src/Ext/RadSite/Hooks.cpp (hunk 5 above)
Comment (NetsuNegi) [reply src/Ext/RadSite/Hooks.cpp:145]: it's structured bindings :)

**Review by Metadorius — APPROVED** [review]
